cmake/OpenCVCompilerOptions.cmake

```diff
@@ -146,8 +146,11 @@ if(CMAKE_COMPILER_IS_GNUCXX)
   elseif(X86 OR X86_64)
     add_extra_compiler_option(-mno-sse2)
   endif()
+  if(ARM)
+    add_extra_compiler_option("-mfp16-format=ieee")
+  endif(ARM)
   if(ENABLE_NEON)
-    add_extra_compiler_option("-mfpu=neon")
+    add_extra_compiler_option("-mfpu=neon-fp16")
```

Comment:
- Compiler support
  - I checked and indeed, some early compiler (gcc 4.4) doesn't accept neon-fp16.  I'll put something to guard this part
- Architecture support
  - Technically, fp16 is not architecture support. It's included in FPU. From ARM v8, it's the part of the architecture.
  - This PR includes the check of feature on run time, so It shouldn't be a problem with older processor.